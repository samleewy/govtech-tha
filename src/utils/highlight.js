// takes in highlight in the form: [{ BeginOffset: X, EndOffset: X }]
// wraps those text with class 'highlight'
export const highlightText = (highlights, text) => {
    let prevOffset = 0;
    if (highlights.length === 0) return text;
    return <span>
        { highlights.map((highlight, idx) => {
            const elements = (
                <div key={idx}>
                    {highlight.BeginOffset != 0 && <span>{ text.slice(prevOffset, highlight.BeginOffset) }</span>}
                    <span className='highlight'>{ text.slice(highlight.BeginOffset, highlight.EndOffset) }</span>
                </div>
            )
            prevOffset = highlight.EndOffset;
            return elements
        }) }
        <span>{ text.slice(prevOffset, text.length) }</span>
    </span>
}

// finds all indices (begin, end) of all terms in text
export const getHighlightIndices = (text, term) => {
    let result, indices = [];
    const regex = new RegExp(term, "gi")
    while ((result = regex.exec(text))) {
        indices.push({
            BeginOffset: result.index,
            EndOffset: result.index + term.length,
        })
    }
    return indices;
}