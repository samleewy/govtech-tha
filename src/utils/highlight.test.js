import { render } from "@testing-library/react";
import { getHighlightIndices, highlightText } from "./highlight";

describe('This suite is to test highlight functions', () => {
    it('should return original text if there are no highlights available', () => {
        const highlights = [];
        const text = "This is a test text";
        const result = highlightText(highlights, text);
        expect(result).toBe(text);
    })
    it('should return correctly highlighted text with class \'highlight\'', () => {
        const highlights = [
            {
                BeginOffset: 0,
                EndOffset: 4,
            },
            {
                BeginOffset: 48,
                EndOffset: 52,
            },
            {
                BeginOffset: 57,
                EndOffset: 61,
            }
        ];
        const text = "test this would highlight everything related to test.... test?"
        const result = highlightText(highlights, text);
        const { asFragment } = render(<>{ result }</>)
        expect(asFragment()).toMatchSnapshot();
    })
    it('should correctly return highlighted text indices', () => {
        const text = "test this would highlight everything related to test.... test?"
        const result = getHighlightIndices(text, 'test')
        const expectedHighlight = [
            {
                BeginOffset: 0,
                EndOffset: 4,
            },
            {
                BeginOffset: 48,
                EndOffset: 52,
            },
            {
                BeginOffset: 57,
                EndOffset: 61,
            }
        ];
        expect(JSON.stringify(result)).toBe(JSON.stringify(expectedHighlight));
    })
})