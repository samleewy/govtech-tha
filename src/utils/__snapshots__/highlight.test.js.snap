// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`This suite is to test highlight functions should return correctly highlighted text with class 'highlight' 1`] = `
<DocumentFragment>
  <span>
    <div
      style="display: inline;"
    >
      <span
        class="highlight"
      >
        test
      </span>
    </div>
    <div
      style="display: inline;"
    >
      <span>
         this would highlight everything related to 
      </span>
      <span
        class="highlight"
      >
        test
      </span>
    </div>
    <div
      style="display: inline;"
    >
      <span>
        .... 
      </span>
      <span
        class="highlight"
      >
        test
      </span>
    </div>
    <span>
      ?
    </span>
  </span>
</DocumentFragment>
`;
